src/java/org/apache/cassandra/cql3/constraints/ColumnConstraints.java

```diff
@@ -0,0 +1,201 @@
+    {
+        this.constraints = constraints;
+    }
+
+    @Override
+    public IVersionedSerializer<ColumnConstraints> serializer()
+    {
+        return serializer;
+    }
+
+    @Override
+    public void appendCqlTo(CqlBuilder builder)
+    {
+        for (ColumnConstraint<?> constraint : constraints)
+            constraint.appendCqlTo(builder);
+    }
+
+    @Override
+    public void evaluate(Class<? extends AbstractType> valueType, Object columnValue) throws ConstraintViolationException
+    {
+        for (ColumnConstraint<?> constraint : constraints)
+            constraint.evaluate(valueType, columnValue);
+    }
+
+    public List<ColumnConstraint<?>> getConstraints()
+    {
+        return constraints;
+    }
+
+    public boolean isEmpty()
+    {
+        return constraints.isEmpty();
+    }
+
+    public int getSize()
+    {
+        return constraints.size();
+    }
+
+    // Checks if there is at least one constraint that will perform checks
+    public boolean hasRelevantConstraints()
+    {
+        for (ColumnConstraint c : constraints)
+        {
+            if (c != ColumnConstraints.NO_OP)
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void validate(ColumnMetadata columnMetadata) throws InvalidConstraintDefinitionException
+    {
+        if (!columnMetadata.type.isConstrainable())
+            throw new InvalidConstraintDefinitionException("Constraint cannot be defined on the column "
+                                                           + columnMetadata.name + " of type " + columnMetadata.type.asCQL3Type()
+                                                           + " for the table " + columnMetadata.ksName + "." + columnMetadata.cfName);
+
+        for (ColumnConstraint<?> constraint : constraints)
+            constraint.validate(columnMetadata);
+    }
+
+    @Override
+    public ConstraintType getConstraintType()
+    {
+        return ConstraintType.COMPOSED;
+    }
+
+    private static class Noop extends ColumnConstraints
+    {
+        private Noop()
+        {
+            super(Collections.emptyList());
+        }
+
+        @Override
+        public void validate(ColumnMetadata columnMetadata)
+        {
+            // Do nothing. It is always valid
+        }
+    }
+
+    public final static class Raw
+    {
+        private final List<ColumnConstraint<?>> constraints;
+
+        public Raw(List<ColumnConstraint<?>> constraints)
+        {
+            this.constraints = constraints;
+        }
+
+        public Raw()
+        {
+            this.constraints = Collections.emptyList();
+        }
+
+        public ColumnConstraints prepare()
+        {
+            if (constraints.isEmpty())
+                return NO_OP;
+            return new ColumnConstraints(constraints);
+        }
+    }
+
+    public static class Serializer implements IVersionedSerializer<ColumnConstraints>
```

Comment:
I think this (and the serializers on `ColumnConstraint` & its impls) should implement `MetadataSerializer` rather than `IVersionedSerializer`. The latter is primarily used for internode messaging, whereas the former is for writing entries in the cluster metadata log and for serializing `ClusterMetadata` snapshots (which are both persisted locally and sent over the wire). Constraints only seem to require serialization as components of schema in a `ClusterMetadata`.